### PR TITLE
[dbnode] Add continuous file profile with RPC endpoint

### PR DESCRIPTION
**Pull request by robskillington**
**What this PR does / why we need it**:

Allows for collecting profiles when a specific condition is triggered./

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note
NONE
```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note
NONE
```

### CODE REVIEW

File: src/x/debug/triggering_profile.go [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+const (
+	// ContinuousCPUProfileName is the name of continuous CPU profile.
+	ContinuousCPUProfileName = "cpu"
+)
+
+var (
+	defaultConditional     = func() bool { return true }
+	defaultInterval        = time.Second
+	errNoFilePathTemplate  = errors.New("no file path template")
+	errProfileName         = errors.New("no profile name")
```
Comment (martin-mao) [inline src/x/debug/triggering_profile.go:46]: Nit: Call this errNoProfileName for consistency.

File: src/x/debug/triggering_profile.go [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+	}
+
+	c.closeCh = make(chan struct{})
+	go c.run()
+	return nil
+}
+
+// Stop will stop the continuous file profile.
+func (c *ContinuousFileProfile) Stop() error {
+	if c.closeCh == nil {
```
Comment (martin-mao) [inline src/x/debug/triggering_profile.go:134]: Does this need a lock for race of multiple start/stop calls?

File: src/x/debug/triggering_profile.go [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+			_ = w.Close()
+		}
+	}()
+
+	switch c.profileName {
+	case ContinuousCPUProfileName:
+		if err := pprof.StartCPUProfile(w); err != nil {
+			return err
+		}
+		time.Sleep(c.profileDuration)
```
Comment (martin-mao) [inline src/x/debug/triggering_profile.go:198]: Is there an issue where the profileDuration is longer than the tick interval and we kick off more than one profile at a time?

File: src/x/debug/triggering_profile.go (hunk 2 above)
Comment (robskillington) [reply src/x/debug/triggering_profile.go:134]: Only mean to be use in a single threaded way, so no problem here.

File: src/x/debug/triggering_profile.go (hunk 3 above)
Comment (robskillington) [reply src/x/debug/triggering_profile.go:198]: Thankfully not, higher level loop waits for this method to finish 👍

File: src/x/debug/triggering_profile.go (hunk 1 above)
Comment (robskillington) [reply src/x/debug/triggering_profile.go:46]: Sg

File: src/x/debug/triggering_profile.go [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+		select {
+		case <-closeCh:
+			return
+		case <-ticker.C:
+			if !c.conditional() {
+				continue
+			}
+
+			err := c.profile()
+			if err != nil {
```
Comment (arnikola) [inline src/x/debug/triggering_profile.go:159]: Should this close channels?

File: src/x/debug/triggering_profile_test.go [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+					if strings.HasSuffix(f.Name(), ext) {
+						require.FailNow(t, "conditional false, profile generated")
+					}
+				}
+			}
+
+			// Trigger profile.
+			cond.Store(true)
+
+			for start := time.Now(); time.Since(start) < 10*time.Second; {
```
Comment (arnikola) [inline src/x/debug/triggering_profile_test.go:97]: 10 second test timeout is on the longer side, how long does this usually take to hit?

**Pull request merged**